File: AspNetMvcUnleashed/0672329980_source/Chapter06/CS/MvcApplication1/MvcApplication1/Helpers/BulletedListHelper.cs
using System;
using System.Collections;
using System.IO;
using System.Web.Mvc;
using System.Web.UI;

namespace Helpers
{
    public static class BulletedListHelper
    {
        public static string BulletedList(this HtmlHelper helper, string name)
        {
            var items = helper.ViewData.Eval(name) as IEnumerable;
            if (items == null)
                throw new NullReferenceException("Cannot find " + name + " in view data");

            var writer = new HtmlTextWriter(new StringWriter());
            
            // Open UL
            writer.RenderBeginTag(HtmlTextWriterTag.Ul);
            foreach (var item in items)
            {
                writer.RenderBeginTag(HtmlTextWriterTag.Li);
                writer.Write(helper.Encode(item));
                writer.RenderEndTag();
                writer.WriteLine();
            }
            // Close UL
            writer.RenderEndTag();

            // Return the HTML string
            return writer.InnerWriter.ToString();
        }
    }
}

File: AspNetMvcUnleashed/0672329980_source/Chapter06/VB/MvcApplication1/MvcApplication1/Helpers/BulletedListHelper.vb
Imports System.IO

Public Module BulletedListHelper

    <System.Runtime.CompilerServices.Extension()> _
    Function BulletedList(ByVal helper As HtmlHelper, ByVal name As String) As String
        Dim items As IEnumerable = helper.ViewData.Eval(name)
        If items Is Nothing Then
            Throw New NullReferenceException("Cannot find " & name & " in view data")
        End If

        Dim writer = New HtmlTextWriter(New StringWriter())

        ' Open UL
        writer.RenderBeginTag(HtmlTextWriterTag.Ul)
        For Each item In items
            writer.RenderBeginTag(HtmlTextWriterTag.Li)
            writer.Write(helper.Encode(item))
            writer.RenderEndTag()
            writer.WriteLine()
        Next item
        ' Close UL
        writer.RenderEndTag()

        ' Return the HTML string
        Return writer.InnerWriter.ToString()
    End Function

End Module
